---
output: pdf_document
header-includes:
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \usepackage{xcolor}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{fancyhdr}
- \pagestyle{fancy}
- \fancyhf{}
- \setlength{\headheight}{23pt}
- \chead{\textbf{Enteric Diseases Laboratory Branch SARS-CoV-2 Variant Whole Genome Sequencing Run Summary}}
- \cfoot{\textbf{Ver. No.}01}
- \lfoot{\textbf{Doc. No.} EDLB.TE.C.032.F02}
- \rfoot{\textbf{Effective Date:}TBD}
classoption:
  - landscape
params:
  analysisDirFP: null
  runID: null
  seqDirFP: null
  sampleIDInsertFreq: 5
  runNormal: TRUE

---

```{r index-comments, include=FALSE, eval=FALSE}
# For rapid testing:
rmarkdown::render(input = "Cecret/bin/report/clia_summary.Rmd", output_file = "clia_summary.pdf", output_format = "pdf_document", params = list(analysisDirFP = "../../../../testing", runID = "M3235-21-015", seqDirFP = "../../../../testing"))

# To generate an empty form for CLIA form sign off (rarely used)
rmarkdown::render(input = "Cecret/bin/report/clia_summary.Rmd", output_file = "clia_summary.pdf", output_format = "pdf_document", params = list(analysisDirFP = "../../../../testing", runID = "M3235-21-015", seqDirFP = "../../../../testing", runNormal = FALSE))
```

```{r setup1, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(knitr)
library(kableExtra)
library(rmarkdown)
```

```{r read-tabs, include=FALSE, eval=params$runNormal}
# Read summary.txt in
sumTab <- read_tsv(file.path(params$analysisDirFP, "summary.txt"))
```

```{r format-sumTab, include=FALSE, eval=params$runNormal}
# Formatting summary.txt.
sumTab <- select(sumTab, -aligner_version, -ivar_version, -fastp_reads_passed) %>%
          mutate(S_aa_INDELs = str_replace_all(S_aa_INDELs, c("," = ", ")))

# Make summary.txt column names nicer
colnames(sumTab) <- c("Sample.ID", 
                      "Sample.Name", 
                      "Pangolin", 
                      "Pangolin.QC", 
                      "NextClade", 
                      "#FastQC.R1", 
                      "#FastQC.R2", 
                      "#Seqyclean.Pairs", 
                      "%Seqyclean.Pairs", 
                      "Depth.Post.Trim", 
                      "Coverage.Post.Trim", 
                      "%Human.Reads", 
                      "%SC2.Reads", 
                      "#iVar.Variants", 
                      "#BCFTools.Variants", 
                      "#BEDTools.Failed.Amps", 
                      "#SAMTools.Failed.Amps", 
                      "#N", 
                      "#Degenerate", 
                      "#ACTG", 
                      "#Total.Bases", 
                      "Total.Reads.Analyzed", 
                      "%N", 
                      "Mean.Cov.Depth", 
                      "%Reads.Mapping.SC2", 
                      "Vadr", 
                      "Vadr.All.ORF.Shift",
                      "Vadr.S.ORF.Shift", 
                      "AA.Changes.S",
                      "Length.Longest.Insert", 
                      "Length.Longest.Del",
                      "pangoLEARN.v",
                      "#Lineage.Subs",
                      "GenBank#",
                      "ORFs.Passing.QC",
                      "Coverage.S",
                      "Mean.Depth.S",
                      "Percent.Pos.Min.Cov.S",
                      "Percent.Ns.S"
)
```

```{r format-sumTab-empty, include=FALSE, eval=ifelse(params$runNormal == FALSE, TRUE, FALSE)}
#Make empty sumTab
sumTab <- tibble(Sample.ID = character(), 
                  Sample.Name = character(), 
                  Pangolin = character(), 
                  Pangolin.QC = character(), 
                  NextClade = character(), 
                  `#FastQC.R1` = numeric(), 
                  `#FastQC.R2` = numeric(), 
                  `#Seqyclean.Pairs` = numeric(), 
                  `%Seqyclean.Pairs` = numeric(), 
                  Depth.Post.Trim = numeric(), 
                  Coverage.Post.Trim = numeric(), 
                  `%Human.Reads` = numeric(), 
                  `%SC2.Reads` = numeric(), 
                  `#iVar.Variants` = numeric(), 
                  `#BCFTools.Variants` = numeric(), 
                  `#BEDTools.Failed.Amps` = numeric(), 
                  `#SAMTools.Failed.Amps` = numeric(), 
                  `#N` = numeric(), 
                  `#Degenerate` = numeric(), 
                  `#ACTG` = numeric(), 
                  `#Total.Bases` = numeric(), 
                  Total.Reads.Analyzed = numeric(), 
                  `%N` = numeric(), 
                  Mean.Cov.Depth = numeric(), 
                  `%Reads.Mapping.SC2` = numeric(), 
                  Vadr = character(), 
                  Vadr.All.ORF.Shift = character(),
                  Vadr.S.ORF.Shift = character(), 
                  AA.Changes.S = character(),
                  Length.Longest.Insert = numeric(), 
                  Length.Longest.Del = numeric(),
                  pangoLEARN.v = character(),
                  `#Lineage.Subs` = numeric(),
                  `GenBank#` = character(),
                  ORFs.Passing.QC = numeric(),
                  Coverage.S = numeric(),
                  Mean.Depth.S = numeric(),
                  Percent.Pos.Min.Cov.S = numeric(),
                  Percent.Ns.S = numeric()
)
```


```{r format-tabs, include=FALSE}
# split out subtables for eLIMS and AA.Changes.S 
aaTab <- select(sumTab, 
                Sample.Name, 
                `Spike Protein Substitutions` = AA.Changes.S) %>%
         arrange(Sample.Name)
elimsTab <- select(sumTab,
                   Sample.Name,
                   Average.Depth = Depth.Post.Trim,
                   Percent.Genome.Coverage = Coverage.Post.Trim,
                   Total.Reads = `#FastQC.R1`,
                   Percent.Mapped.Reads = `%Reads.Mapping.SC2`,
                   Open.Reading.Frames = ORFs.Passing.QC,
                   S.Gene.Coverage = Coverage.S,
                   Lineage = Pangolin,
                   Number.Lineage.Defined.Substitutions = `#Lineage.Subs`,
                   pangoLEARN.version = pangoLEARN.v,
                   `GenBank.Accession#` = `GenBank#`) %>%
            arrange(Sample.Name)
qcTab <- select(sumTab,
                Sample.Name,
                Pangolin.QC,
                `%Ambiguous.Base.Calls` = `%N`,
                `%Ambiguous.Base.Calls.S` = Percent.Ns.S,
                Frameshifts = Vadr.S.ORF.Shift) %>%
          arrange(Sample.Name)
sumTab <- select(sumTab,
                 -AA.Changes.S,
                 -Depth.Post.Trim,
                 -Coverage.Post.Trim,
                 -`#FastQC.R1`,
                 -`%Reads.Mapping.SC2`,
                 -ORFs.Passing.QC,
                 -Coverage.S,
                 -Pangolin,
                 -`#Lineage.Subs`,
                 -pangoLEARN.v,
                 -`GenBank#`,
                 -Pangolin.QC,
                 -`%N`,
                 -Percent.Ns.S,
                 -Vadr.S.ORF.Shift) %>%
          arrange(Sample.Name)

# Round off some numbers to make them prettier
elimsTab <- mutate(elimsTab,
                   Percent.Mapped.Reads = round(Percent.Mapped.Reads, digits = 1),
                   S.Gene.Coverage = round(S.Gene.Coverage, digits = 1),
                   Average.Depth = round(Average.Depth, digits = 1),
                   Percent.Genome.Coverage = round(Percent.Genome.Coverage, digits = 1))
qcTab <- mutate(qcTab,
                `%Ambiguous.Base.Calls` = round(`%Ambiguous.Base.Calls`, digits = 1),
                `%Ambiguous.Base.Calls.S` = round(`%Ambiguous.Base.Calls.S`, digits = 1))
sumTab <- mutate(sumTab,
                 `%Seqyclean.Pairs` = round(`%Seqyclean.Pairs`, digits = 1),
                 `%SC2.Reads` = round(`%SC2.Reads`, digits = 1),
                 Mean.Cov.Depth = round(Mean.Cov.Depth, digits = 1))
# # 
# # Insert SampleID column into sumTab at every nth column (params$sampleIDInsertFreq) position
# # how many times to insert sampleID column
# colRepsSumTab <- floor(ncol(sumTab) / params$sampleIDInsertFreq) 
# colRepsElimsTab <- floor(ncol(elimsTab) / params$sampleIDInsertFreq) 
# # create and bind columns to insert
# sumTab <- bind_cols(sumTab, as_tibble(replicate(colRepsSumTab, sumTab$Sample.Name))) %>%
#           relocate(Sample.Name)
# sumTabColNames <- colnames(sumTab)
# elimsTab <- bind_cols(elimsTab, as_tibble(replicate(colRepsElimsTab, elimsTab$Sample.Name))) 
# elimsTabColNames <- colnames(elimsTab)
# 
# # function to rename columns slightly more nicely. Note R won't allow redundant column names.
# colNamer <- function(c) {
#   return(paste0("Sample.Name", str_replace(c, "V", "")))
# }
# # move new sampleID columns to right place
# for (i in 1:colRepsSumTab) {
#   sumTab <- relocate(sumTab, 
#                      eval(paste0("V", i)), 
#                      .after = sumTabColNames[i*params$sampleIDInsertFreq]) %>%
#             rename_with(colNamer, starts_with(paste0("V", i)))
# }
# for (i in 1:colRepsElimsTab) {
#   elimsTab <- relocate(elimsTab, 
#                      eval(paste0("V", i)), 
#                      .after = elimsTabColNames[i*params$sampleIDInsertFreq]) %>%
#             rename_with(colNamer, starts_with(paste0("V", i)))
# }

```

\begin{center}
\textbf{\huge{eLIMS Fields}}
\end{center}

```{r display-tabs, echo=FALSE, results='asis'}

# myTableSplittr <- function(t) {
#   # Where t is a table that is too wide for one page and needs to be split up
#   # returns nothing; run for side effect of printing tables
#   sampleIDIndices <- which(str_starts(colnames(t), "Sample.Name"))
#   for (s in 1:length(sampleIDIndices)) {
#     if (sampleIDIndices[s] == head(sampleIDIndices, n = 1)) {
#       i <- sampleIDIndices[s] + as.numeric(params$sampleIDInsertFreq) - 1
#       y <- colnames(t)[1:i]
#     } else {
#       if (sampleIDIndices[s] < tail(sampleIDIndices, n = 1)) {
#         i <- sampleIDIndices[s] + as.numeric(params$sampleIDInsertFreq)
#         y <- colnames(t)[sampleIDIndices[s]:i]
#       } else {
#         i <- length(colnames(t))
#         y <- colnames(t)[sampleIDIndices[s]:i]
#       }
#     }
#     if (length(y) > 1) {
#       print(kbl(select(t, all_of(y)),
#                 booktabs = T,
#                 longtable = T,
#                 align = "l") %>%
#             kable_styling(latex_options = c("striped", "repeat_header")) %>%
#             column_spec(1, width = "15em"))
#       cat('\\pagebreak')
#     }
#   }
#   return(cat('\\pagebreak'))
# }

newTableSplittr <- function(t, f, c) {
  # Where t is a tibble that is too wide to display on one page
  # Where f is the frequency with which to split
  # Where c is the column name to repeat at the beginning of each split
  repeatCol <- select(t, matches(c))
  t <- select(t, -matches(c))
  myColNames <- colnames(t)
  numSplits <- ceiling(length(myColNames) / f)
  for (g in 1:numSplits) {
    x <- ifelse(g * f <= length(myColNames), g * f, length(myColNames))
    subT <- select(t, all_of(myColNames[(g * f - f + 1):x])) %>%
      bind_cols(repeatCol) %>%
      relocate(Sample.Name)
    print(kbl(subT,
              booktabs = T,
              longtable = T,
              align = "l") %>%
          kable_styling(latex_options = c("striped", "repeat_header")) %>%
          column_spec(1, width = "15em"))
    cat('\\pagebreak')
  }
  return(cat('\\pagebreak'))
}

newTableSplittr(elimsTab, params$sampleIDInsertFreq, "Sample.Name")

print(kbl(aaTab,
          booktabs = T,
          longtable = T,
          align = "l") %>%
      kable_styling(latex_options = c("striped", "repeat_header")) %>%
      column_spec(1, width = "15em") %>%
      column_spec(2, width = "45em"))
cat('\\pagebreak')
```

\begin{center}
\textbf{\huge{Other QC Fields}}
\end{center}

```{r display-qcTab, echo=FALSE, results='asis'}
print(kbl(qcTab,
          booktabs = T,
          longtable = T,
          align = "l") %>%
      kable_styling(latex_options = c("striped", "repeat_header")) %>%
      column_spec(1, width = "15em"))
cat('\\pagebreak')
```

\begin{center}
\textbf{\huge{Other Dry Lab Fields}}
\end{center}

```{r display-sumTab, echo=FALSE, results='asis'}
newTableSplittr(sumTab, params$sampleIDInsertFreq, "Sample.Name")
```
