---
output: pdf_document
header-includes:
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \usepackage{xcolor}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{fancyhdr}
- \pagestyle{fancy}
- \fancyhf{}
- \setlength{\headheight}{23pt}
- \chead{\textbf{Enteric Diseases Laboratory Branch SARS-CoV-2 Variant Whole Genome Sequencing Run Summary}}
- \cfoot{\textbf{Ver. No.}01}
- \lfoot{\textbf{Doc. No.} EDLB.TE.C.032.F02}
- \rfoot{\textbf{Effective Date:}TBD}
classoption:
  - landscape
params:
  analysisDirFP: null
  runID: null
  seqDirFP: null
  sampleIDInsertFreq: 5

---

```{r index-comments, include=FALSE, eval=FALSE}
# For rapid testing:
rmarkdown::render(input = "Cecret/bin/report/clia_summary.Rmd", output_file = "clia_summary.pdf", output_format = "pdf_document", params = list(analysisDirFP = "../../../../testing", runID = "M3235-21-015", seqDirFP = "../../../../testing"))
```

```{r setup1, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
#library(formattable)
library(knitr)
library(kableExtra)
library(rmarkdown)
```

```{css, echo=FALSE}
h1, h2, h3 {
  text-align: center;
}
```

```{r read-tabs, include=FALSE}
# Read summary.txt in
sumTab <- read_tsv(file.path(params$analysisDirFP, "summary.txt"))
```

```{r format-tabs, include=FALSE}
# Formatting summary.txt.
sumTab <- select(sumTab, -aligner_version, -ivar_version, -fastp_reads_passed) %>%
          mutate(S_aa_INDELs = str_replace_all(S_aa_INDELs, c("," = ", ")))

# Make summary.txt column names nicer
colnames(sumTab) <- c("Sample.ID", 
                      "Sample.Name", 
                      "Pangolin", 
                      "Pangolin.QC", 
                      "NextClade", 
                      "#FastQC.R1", 
                      "#FastQC.R2", 
                      "#Seqyclean.Pairs", 
                      "%Seqyclean.Pairs", 
                      "Depth.Post.Trim", 
                      "Coverage.Post.Trim", 
                      "%Human.Reads", 
                      "%SC2.Reads", 
                      "#iVar.Variants", 
                      "#BCFTools.Variants", 
                      "#BEDTools.Failed.Amps", 
                      "#SAMTools.Failed.Amps", 
                      "#N", 
                      "#Degenerate", 
                      "#ACTG", 
                      "#Total.Bases", 
                      "Total.Reads.Analyzed", 
                      "%N", 
                      "Mean.Cov.Depth", 
                      "%Reads.Mapping.SC2", 
                      "Vadr", 
                      "Vadr.All.ORF.Shift",
                      "Vadr.S.ORF.Shift", 
                      "AA.Changes.S",
                      "Length.Longest.Insert", 
                      "Length.Longest.Del",
                      "pangoLEARN.v",
                      "#Lineage.Subs",
                      "Percent.Mapped.Reads",
                      "GenBank#",
                      "ORFs.Passing.QC",
                      "Coverage.S",
                      "Mean.Depth.S",
                      "Percent.Pos.Min.Cov.S",
                      "Percent.Ns.S"
)

# split out subtables for eLIMS and AA.Changes.S 
aaTab <- select(sumTab, 
                Sample.Name, 
                `Spike Protein Substitutions` = AA.Changes.S) %>%
         arrange(Sample.Name)
elimsTab <- select(sumTab,
                   Sample.Name,
                   Lineage = Pangolin,
                   Total.Reads = `#FastQC.R1`,
                   Average.Depth = Depth.Post.Trim,
                   Percent.Genome.Coverage = Coverage.Post.Trim,
                   Mapped.Reads = Total.Reads.Analyzed,
                   Open.Reading.Frames = ORFs.Passing.QC,
                   S.Gene.Coverage = Coverage.S,
                   Percent.Mapped.Reads,
                   Number.Lineage.Defined.Substitutions = `#Lineage.Subs`,
                   pangoLEARN.version = pangoLEARN.v,
                   `GenBank.Accession#` = `GenBank#`) %>%
            arrange(Sample.Name)
sumTab <- select(sumTab,
                 -AA.Changes.S,
                 -Pangolin,
                 -`#FastQC.R1`,
                 -Depth.Post.Trim,
                 -Coverage.Post.Trim,
                 -Total.Reads.Analyzed,
                 -ORFs.Passing.QC,
                 -Coverage.S,
                 -Percent.Mapped.Reads,
                 -`#Lineage.Subs`,
                 -pangoLEARN.v,
                 -`GenBank#`) %>%
          arrange(Sample.Name)

# Insert SampleID column into sumTab at every nth column (params$sampleIDInsertFreq) position
# how many times to insert sampleID column
colRepsSumTab <- floor(ncol(sumTab) / params$sampleIDInsertFreq) 
colRepsElimsTab <- floor(ncol(elimsTab) / params$sampleIDInsertFreq) 
# create and bind columns to insert
sumTab <- bind_cols(sumTab, as_tibble(replicate(colRepsSumTab, sumTab$Sample.Name))) %>%
          relocate(Sample.Name)
sumTabColNames <- colnames(sumTab)
elimsTab <- bind_cols(elimsTab, as_tibble(replicate(colRepsElimsTab, elimsTab$Sample.Name))) 
elimsTabColNames <- colnames(elimsTab)

# function to rename columns slightly more nicely. Note R won't allow redundant column names.
colNamer <- function(c) {
  return(paste0("Sample.Name", str_replace(c, "V", "")))
}
# move new sampleID columns to right place
for (i in 1:colRepsSumTab) {
  sumTab <- relocate(sumTab, 
                     eval(paste0("V", i)), 
                     .after = sumTabColNames[i*params$sampleIDInsertFreq]) %>%
            rename_with(colNamer, starts_with(paste0("V", i)))
}
for (i in 1:colRepsElimsTab) {
  elimsTab <- relocate(elimsTab, 
                     eval(paste0("V", i)), 
                     .after = elimsTabColNames[i*params$sampleIDInsertFreq]) %>%
            rename_with(colNamer, starts_with(paste0("V", i)))
}

```

### eLIMS Fields

```{r display-tabs, echo=FALSE, results='asis'}

myTableSplittr <- function(t) {
  # Where t is a table that is too wide for one page and needs to be split up
  # returns nothing; run for side effect of printing tables
  sampleIDIndices <- which(str_starts(colnames(t), "Sample.Name"))
  for (s in 1:length(sampleIDIndices)) {
    if (sampleIDIndices[s] == head(sampleIDIndices, n = 1)) {
      i <- sampleIDIndices[s] + as.numeric(params$sampleIDInsertFreq) - 1
      y <- colnames(t)[1:i]
    } else {
      if (sampleIDIndices[s] < tail(sampleIDIndices, n = 1)) {
        i <- sampleIDIndices[s] + as.numeric(params$sampleIDInsertFreq)
        y <- colnames(t)[sampleIDIndices[s]:i]
      } else {
        i <- length(colnames(t))
        y <- colnames(t)[sampleIDIndices[s]:i]
      }
    }
    if (length(y) > 1) {
      print(kbl(select(t, all_of(y)),
                booktabs = T,
                longtable = T,
                align = "l") %>%
            kable_styling(latex_options = c("striped", "repeat_header")) %>%
            column_spec(1, width = "15em"))
      cat('\\pagebreak')
    }
  }
  return(cat('\\pagebreak'))
}

myTableSplittr(elimsTab)

print(kbl(aaTab,
          booktabs = T,
          longtable = T,
          align = "l") %>%
      kable_styling(latex_options = c("striped", "repeat_header")) %>%
      column_spec(1, width = "15em") %>%
      column_spec(2, width = "45em"))
cat('\\pagebreak')
```

### Other Dry Lab Fields

```{r display-sumTab, echo=FALSE, results='asis'}
myTableSplittr(sumTab)
```
