---
output: pdf_document
header-includes:
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \usepackage{xcolor}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{fancyhdr}
- \pagestyle{fancy}
- \fancyhf{}
- \setlength{\headheight}{23pt}
- \chead{\textbf{Enteric Diseases Laboratory Branch SARS-CoV-2 Variant Whole Genome Sequencing Run Summary}}
- \cfoot{\textbf{Ver. No.}01}
- \lfoot{\textbf{Doc. No.} EDLB.TE.C.032.F02}
- \rfoot{\textbf{Effective Date:}TBD}
---

```{r index-comments, include=FALSE, eval=FALSE}
# For rapid testing:
rmarkdown::render(input = "Cecret/bin/report/clia_sig_page.Rmd", output_format = "pdf_document", params = list(analysisDirFP = "../../../../testing", runID = "M3235-21-015", seqDirFP = "../../../../testing"))
```

```{r setup2, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(rmarkdown)
```
Reviewed by: _____________________________________   

Date: ________________  
