---
title: "Summary"
params:
  analysisDirFP: null
  runID: null
  seqDirFP: null
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(formattable)
library(kableExtra)
```
```{r thresholds, include=FALSE}
fastqcRawThreshold <- 100000          # min number of raw reads accepted in fastqc_raw column
seqycleanKeptReviewThreshold <- 90    # min percent of reads remaining after seqyclean (pass/review)
seqycleanKeptFailThreshold <- 70      # min percent of reads remaining after seqyclean (review/fail)
depthAfterTrimThreshold <- 100        # min depth after trimming
covAfterTrimThreshold <- 95           # min coverage after trimming
perHumanThreshold <- NULL             # max percentage human reads allowed per sample (not in use)
perSC2Threshold <- 65                 # min percentage SC2 reads allowed per sample
perNThreshold <- 10                   # max percentage N allowed in genome
```

```{r format-fnc, include=FALSE}
highGood <- function(v, t) {
  # accepts value and threshhold and returns green if value >= threshhold and gray if not
  return(ifelse(v >= t, "lightgreen", "lightgray"))
}
```


```{r read-sum, include=FALSE}
# Read summary.txt in
sumTab <- read_tsv(paste0(params$analysisDirFP, "/summary.txt"))

# This does all of the formatting for the table read in above and displayed in the next code chunk. It contains a mix of dplyr, formattable, and kableExtra commands.
sumTab <- mutate(sumTab, sample = str_replace(sample, paste0("-", params$runID), "")) %>% 
          # remember to mod column names to match if columns included change
          select(-aligner_version, -ivar_version, -fastp_reads_passed) %>%
          mutate(
            pangolin_lineage = cell_spec(pangolin_lineage, 
                                         color = ifelse(pangolin_lineage == "None", "red", "black"), 
                                         bold = ifelse(pangolin_lineage == "None", T, F)),
            pangolin_status = cell_spec(pangolin_status, 
                                        color = ifelse(pangolin_status == "passed_qc", "black", "red"), 
                                        bold = ifelse(pangolin_status == "passed_qc", F, T)),
            fastqc_raw_reads_1 = color_bar(color = highGood(fastqc_raw_reads_1, fastqcRawThreshold))(fastqc_raw_reads_1), 
            fastqc_raw_reads_2 = color_bar(color = highGood(fastqc_raw_reads_2, fastqcRawThreshold))(fastqc_raw_reads_2),
            seqyclean_percent_kept_after_cleaning = cell_spec(seqyclean_percent_kept_after_cleaning, 
                                        color = ifelse(seqyclean_percent_kept_after_cleaning < seqycleanKeptFailThreshold, "red", elseif(seqyclean_percent_kept_after_cleaning < seqycleanKeptReviewThreshold, "yellow", "black")), 
                                        bold = ifelse(seqyclean_percent_kept_after_cleaning < seqycleanKeptFailThreshold, T, elseif(seqyclean_percent_kept_after_cleaning < seqycleanKeptReviewThreshold, T, F))),
            depth_after_trimming = cell_spec(depth_after_trimming, 
                                             color = ifelse(depth_after_trimming >= depthAfterTrimThreshold, "black", "red"), 
                                             bold = ifelse(depth_after_trimming >= depthAfterTrimThreshold, F, T)),
            coverage_after_trimming = cell_spec(coverage_after_trimming, 
                                                color = ifelse(coverage_after_trimming >= covAfterTrimThreshold, "black", "red"), 
                                                bold = ifelse(coverage_after_trimming >= covAfterTrimThreshold, F, T)),
            # `%_human_reads` = cell_spec(`%_human_reads`, 
            #                                     color = ifelse(`%_human_reads` < perHumanThreshold, "black", "red"), 
            #                                     bold = ifelse(`%_human_reads` < perHumanThreshold, F, T)),
            `%_SARS-COV-2_reads` = cell_spec(`%_SARS-COV-2_reads`, 
                                                color = ifelse(`%_SARS-COV-2_reads` >= perSC2Threshold, "black", "red"), 
                                                bold = ifelse(`%_SARS-COV-2_reads` >= perSC2Threshold, F, T)),
            `%_N` = cell_spec(`%_N`, 
                              color = ifelse(`%_N` >= perNThreshold, "red", "black"), 
                              bold = ifelse(`%_N` >= perNThreshold, T, F)),
            vadr_status = cell_spec(vadr_status, 
                                     color = ifelse(vadr_status == "PASS", "black", "red"), 
                                     bold = ifelse(vadr_status == "PASS", F, T)))

# Make column names nicer
colnames(sumTab) <- c("GroupID", "SampleID", "Pangolin", "Pangolin.QC", "NextClade", "#FastQC.R1", "#FastQC.R2", "#Seqyclean.Pairs", "%Seqyclean.Pairs", "Depth.Post.Trim", "Coverage.Post.Trim", "%Human.Reads", "%SC2.Reads", "#iVar.Variants", "#BCFTools.Variants", "#BEDTools.Failed.Amps", "#SAMTools.Failed.Amps", "#N", "#Degenerage", "#ACTG", "#Total.Bases", "Total.Reads.Analyzed", "%N", "Mean.Cov.Depth", "%Reads.Mapping.SC2", "Vadr", "Length.Longest.Insert", "Length.Longest.Del")
          
```

```{r display-sum, echo=FALSE}
kbl(sumTab, escape = F, align = "l") %>% 
  kable_material(c("striped", "hover")) 
```

